File: 프로그래머스/알고리즘 고득점 Kit/최소직사각형.java
import java.util.*;

class Solution {

	public int solution(int[][] sizes) {
		final List<BusinessCard> list = new ArrayList<>();

		Arrays.stream(sizes).forEach(it ->
			list.add(new BusinessCard(it))
		);

		final BusinessCard maxOfMax = list.stream()
			.max((it1, it2) -> it1.getMax() - it2.getMax())
			.get();
		final BusinessCard maxOfMin = list.stream()
			.max((it1, it2) -> it1.getMin() - it2.getMin())
			.get();

		return maxOfMax.getMax() * maxOfMin.getMin();
	}
}

class BusinessCard {

	private int width;
	private int length;

	public BusinessCard(int[] arr) {
		this.width = arr[0];
		this.length = arr[1];
	}

	public int getMax() {
		return Math.max(this.width, this.length);
	}

	public int getMin() {
		return Math.min(this.width, this.length);
	}
}

File: 프로그래머스/알고리즘 고득점 Kit/최소직사각형.kt
class Solution {
    fun solution(sizes: Array<IntArray>): Int {
        val list = sizes.map { BusinessCard(it) }.toList()

        val maxOfMax = list.maxOf { it.max }
        val maxOfMin = list.maxOf { it.min }

        return maxOfMax * maxOfMin
    }
}

class BusinessCard(
    private val width: Int,
    private val length: Int
) {

    constructor(arr: IntArray): this(arr[0], arr[1])

    val max: Int
        get() {
            return this.width.coerceAtLeast(this.length)
        }

    val min: Int
        get() {
            return this.width.coerceAtMost(this.length)
        }
}
